src/domain/mongoose_domain_loader.erl

```diff
@@ -27,44 +28,50 @@ load_data_from_base_loop(FromId, PageSize) ->
             load_data_from_base_loop(PageMaxId, PageSize)
     end.
 
+remove_outdated_domains_from_core() ->
+    OutdatedDomains = mongoose_domain_core:get_all_outdated(),
+    remove_domains(OutdatedDomains).
+
 check_for_updates(FromId, PageSize) ->
-    check_if_from_num_still_relevant(FromId),
-    %% Ordered by the earlist events first
+    %% Ordered by the earliest events first
     try mongoose_domain_sql:select_updates_from(FromId, PageSize) of
-        [] -> FromId;
         Rows ->
-            PageMaxId = row_to_id(lists:last(Rows)),
-            apply_changes(Rows),
-            check_for_updates(PageMaxId, PageSize)
+            case check_if_id_is_still_relevant(FromId, Rows) of
+                [] -> FromId;
+                RowsToApply ->
+                    PageMaxId = row_to_id(lists:last(RowsToApply)),
+                    apply_changes(RowsToApply),
+                    check_for_updates(PageMaxId, PageSize)
+            end
     catch Class:Reason:StackTrace ->
-            %% Don't allow to crash in the critical code,
-            %% once we've started.
-            ?LOG_ERROR(#{what => domain_check_for_updates_failed,
-                         class => Class, reason => Reason,
-                         stacktrace => StackTrace}),
-            FromId
+        %% Don't allow to crash in the critical code,
+        %% once we've started.
+        ?LOG_ERROR(#{what => domain_check_for_updates_failed,
+                     class => Class, reason => Reason,
+                     stacktrace => StackTrace}),
+        FromId
     end.
 
-%% Be aware that for this check to work, the cleaner should keep at least
-%% one record in domain_events table.
-check_if_from_num_still_relevant(0) ->
-    ok;
-check_if_from_num_still_relevant(FromId) ->
-    Min = mongoose_domain_sql:get_min_event_id(),
-    if Min =:= 0 ->
-            %% Nothing to do, there were no updates done ever in the DB
-            ok;
-       Min > FromId ->
-           %% Looks like this node has no DB connection for a long time.
-           %% But the event log in the DB has been truncated by some other node
-           %% meanwhile. We have to load the whole set of data from DB.
-           Text = <<"DB domain log had some updates to domains deleted, "
-                    " which we have not applied yet. Have to crash.">>,
-           ?LOG_CRITICAL(#{what => events_log_out_of_sync,
-                           text => Text, min_db => Min, from_id => FromId}),
-           mongoose_domain_utils:halt_node(Text);
-       true ->
-           ok
+
+check_if_id_is_still_relevant(FromId, Rows) ->
+    MinId = row_to_id(hd(Rows)),
+    if
+        FromId =:= MinId ->
+            tl(Rows);
+        FromId =:= MinId - 1 ->
+            %% looks like someone completely erased the events table
+            %% this should not happen, but we are still fine.
+            Rows;
+        true ->
```

Comment:
Just a note: this covers both `FromId < MinId - 1` and `FromId > MinId`. I think it is ok to crash in both cases, but the error message is a bit irrelevant in the second one, which should not happen.

Reply:
Maybe it's not obvious from the code of this function, but `FromId > MinId` case is impossible because we query data (in the caller function) starting from `FromID`, see  [check_for_updates/2 ](https://github.com/esl/MongooseIM/pull/3061/files/4e5aad046d69610a4ffd351163ccbc191aee6d3f#diff-0ce59ca69aebb5afbe5caf8b74f0bca372211b36840261e527deb4864661fc7dR37) function.

Reply:
It is clear now, what if we keep `FromId < MinId - 1` instead of `true` there to make it less confusing for the reader? Just a suggestion, not a must.